### PR TITLE
Add FXIOS-7408 [v122] Implement `UILargeContentViewer` and related protocols for URL bar, tab bar and tab location view

### PR DESCRIPTION
**Pull request by atfelix**
## :scroll: Tickets

[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7408)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/16422)

## :bulb: Description

I've added large text support for the buttons that can't grow in size on the tab bar (on the phone in portrait orientation), the tab location view (iPad orientations and other phone orientations) and the URL Bar view (iPad orientations and other phone orientations).

There are several buttons to test:

* Back button
* Forward button
* Multistate button (Home / Search / Reload)
* Tabs button
* Menu button
* Add new tab button
* Search button
* Stop button
* Close tab button (on iPad)
* Bookmarks button
* Share button
* Tracking Protection button (lock icon)
* Reload button
* Privacy mode button
* Reader view button
* QR Code button
* Back button when focused on the search bar
* Search engine button

Note that some of these button have long press gesture recognizers attached.  For the large content viewer to display and the long press gesture recognizers to work as expected, `TabToolbarHelper` now holds a reference to the gesture recognizers and implements the `UIGestureRecognizerDelegate`.  When long pressing on these buttons, you should initially see the large content viewer and then the original long press gesture should activate.  Note the original long press will only work if your touch doesn't leave the original button's view.

In a few places, I've reused an existing accessibility label as a large content title.  

___Question___:  Should I create new localized strings instead of reusing an existing accessibility label

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [x] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [x] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

File: Client/Frontend/Browser/TabTrayButtonExtensions.swift [hunk 1]
```diff
@@ -13,6 +13,9 @@ class PrivateModeButton: ToggleButton, PrivateModeUI {
         accessibilityHint = .TabTrayToggleAccessibilityHint
         let maskImage = UIImage(named: ImageIdentifiers.privateMaskSmall)?.withRenderingMode(.alwaysTemplate)
         setImage(maskImage, for: [])
+        showsLargeContentViewer = true
+        largeContentTitle = .TabTrayToggleAccessibilityLabel
+        largeContentImage = maskImage
```
Comment (atfelix) [inline Client/Frontend/Browser/TabTrayButtonExtensions.swift:18]: I reused an existing accessibility label.  Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Browser/TopTabCell.swift [hunk 2]
```diff
@@ -65,10 +65,16 @@ class TopTabCell: UICollectionViewCell, ThemeApplicable, LegacyTabTrayCell, Reus
...
         titleText.text = tab.getTabTrayTitle()
         accessibilityLabel = getA11yTitleLabel(tab: tab)
+        showsLargeContentViewer = true
+        largeContentTitle = tab.getTabTrayTitle()
         isAccessibilityElement = true
 
         closeButton.accessibilityLabel = String(format: .TopSitesRemoveButtonAccessibilityLabel,
                                                 self.titleText.text ?? "")
+        closeButton.showsLargeContentViewer = true
+        closeButton.largeContentTitle = .TopSitesRemoveButtonLargeContentTitle
```
Comment (atfelix) [inline Client/Frontend/Browser/TopTabCell.swift:75]: I reused an existing accessibility label. Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Browser/TopTabCell.swift [hunk 3]
```diff
@@ -65,10 +65,16 @@ class TopTabCell: UICollectionViewCell, ThemeApplicable, LegacyTabTrayCell, Reus
...
+        closeButton.largeContentImage = UIImage.templateImageNamed(StandardImageIdentifiers.Large.cross)
+        closeButton.scalesLargeContentImage = true
```
Comment (atfelix) [inline Client/Frontend/Browser/TopTabCell.swift:77]: I reused an existing accessibility label. Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Browser/TopTabsViewController.swift [hunk 4]
```diff
@@ -57,13 +57,17 @@ class TopTabsViewController: UIViewController, Themeable, Notifiable {
...
     }
 
     private lazy var newTab: UIButton = .build { button in
         button.setImage(UIImage.templateImageNamed(StandardImageIdentifiers.Large.plus), for: .normal)
         button.semanticContentAttribute = .forceLeftToRight
         button.addTarget(self, action: #selector(TopTabsViewController.newTabTapped), for: .touchUpInside)
         button.accessibilityIdentifier = AccessibilityIdentifiers.Toolbar.addNewTabButton
+        button.accessibilityLabel = .AddTabAccessibilityLabel
+        button.showsLargeContentViewer = true
+        button.largeContentTitle = .AddTabAccessibilityLabel
```
Comment (atfelix) [inline Client/Frontend/Browser/TopTabsViewController.swift:70]: Here I added a missing accessibility label with an existing accessibility label and that  accessibility label. Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Toolbar+URLBar/TabLocationView.swift [hunk 5]
```diff
@@ -102,6 +102,10 @@ class TabLocationView: UIView, FeatureFlaggable {
         trackingProtectionButton.addTarget(self, action: #selector(self.didPressTPShieldButton(_:)), for: .touchUpInside)
         trackingProtectionButton.clipsToBounds = false
         trackingProtectionButton.accessibilityIdentifier = AccessibilityIdentifiers.Toolbar.trackingProtection
+        trackingProtectionButton.showsLargeContentViewer = true
+        trackingProtectionButton.largeContentImage = .templateImageNamed(StandardImageIdentifiers.Large.lock)
+        trackingProtectionButton.largeContentTitle = .TabLocationLockButtonLargeContentTitle
+        trackingProtectionButton.accessibilityLabel = .TabLocationLockButtonAccessibilityLabel
```
Comment (atfelix) [inline Client/Frontend/Toolbar+URLBar/TabLocationView.swift:108]: I reused an existing accessibility label. Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Toolbar+URLBar/TabLocationView.swift [hunk 6]
```diff
@@ -110,15 +114,23 @@ class TabLocationView: UIView, FeatureFlaggable {
         shareButton.contentHorizontalAlignment = .center
         shareButton.accessibilityIdentifier = AccessibilityIdentifiers.Toolbar.shareButton
         shareButton.accessibilityLabel = .TabLocationShareAccessibilityLabel
+        shareButton.showsLargeContentViewer = true
+        shareButton.largeContentImage = .templateImageNamed(ImageIdentifiers.share)
+        shareButton.largeContentTitle = .TabLocationShareButtonAccessibilityLabel
```
Comment (atfelix) [inline Client/Frontend/Toolbar+URLBar/TabLocationView.swift:119]: I reused an existing accessibility label. Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Toolbar+URLBar/TabLocationView.swift [hunk 7]
```diff
@@ -110,15 +114,23 @@ class TabLocationView: UIView, FeatureFlaggable {
...
         button.addTarget(self, action: #selector(self.didPressShoppingButton(_:)), for: .touchUpInside)
         button.isHidden = true
-        button.setImage(UIImage(named: StandardImageIdentifiers.Large.shopping)?.withRenderingMode(.alwaysTemplate), for: .normal)
+        button.setImage(image?.withRenderingMode(.alwaysTemplate), for: .normal)
         button.imageView?.contentMode = .scaleAspectFit
         button.accessibilityLabel = .TabLocationShoppingAccessibilityLabel
         button.accessibilityIdentifier = AccessibilityIdentifiers.Toolbar.shoppingButton
+        button.showsLargeContentViewer = true
+        button.largeContentTitle = .TabLocationShoppingAccessibilityLabel
+        button.largeContentImage = image
```
Comment (atfelix) [inline Client/Frontend/Toolbar+URLBar/TabLocationView.swift:133]: I reused an existing accessibility label. Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Toolbar+URLBar/TabLocationView.swift [hunk 8]
```diff
@@ -135,6 +147,9 @@ class TabLocationView: UIView, FeatureFlaggable {
                 name: .TabLocationReaderModeAddToReadingListAccessibilityLabel,
                 target: self,
                 selector: #selector(self.readerModeCustomAction))]
+        readerModeButton.showsLargeContentViewer = true
+        readerModeButton.largeContentTitle = .TabLocationReaderModeAccessibilityLabel
+        readerModeButton.largeContentImage = .templateImageNamed("reader")
```
Comment (atfelix) [inline Client/Frontend/Toolbar+URLBar/TabLocationView.swift:152]: I reused an existing accessibility label. Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Toolbar+URLBar/TabToolbarHelper.swift [hunk 9]
```diff
@@ -68,20 +69,28 @@ open class TabToolbarHelper: NSObject {
...
             toolbar.multiStateButton.setImage(ImageStop, for: .normal)
             toolbar.multiStateButton.accessibilityLabel = .TabToolbarStopAccessibilityLabel
+            toolbar.multiStateButton.largeContentTitle = .TabToolbarStopAccessibilityLabel
+            toolbar.multiStateButton.largeContentImage = ImageStop
             toolbar.multiStateButton.accessibilityIdentifier = AccessibilityIdentifiers.Toolbar.stopButton
         default:
             toolbar.multiStateButton.setImage(ImageHome, for: .normal)
             toolbar.multiStateButton.accessibilityLabel = .TabToolbarHomeAccessibilityLabel
+            toolbar.multiStateButton.largeContentImage = ImageHome
+            toolbar.multiStateButton.largeContentTitle = .TabToolbarHomeAccessibilityLabel
```
Comment (atfelix) [inline Client/Frontend/Toolbar+URLBar/TabToolbarHelper.swift:93]: I reused existing accessibility labels. Should I create a separate localized strings for the `largeContentTitle`s?

File: Client/Frontend/Toolbar+URLBar/TabToolbarHelper.swift [hunk 10]
```diff
@@ -90,14 +99,22 @@ open class TabToolbarHelper: NSObject {
...
         self.toolbar = toolbar
         super.init()
 
+        toolbar.addUILargeContentViewInteraction(interaction: uiLargeContentViewInteraction)
+
         toolbar.backButton.setImage(UIImage.templateImageNamed(StandardImageIdentifiers.Large.back)?.imageFlippedForRightToLeftLayoutDirection(), for: .normal)
         toolbar.backButton.accessibilityLabel = .TabToolbarBackAccessibilityLabel
         toolbar.backButton.accessibilityIdentifier = AccessibilityIdentifiers.Toolbar.backButton
+        toolbar.backButton.showsLargeContentViewer = true
+        toolbar.backButton.largeContentTitle = .TabToolbarBackAccessibilityLabel
```
Comment (atfelix) [inline Client/Frontend/Toolbar+URLBar/TabToolbarHelper.swift:115]: I reused an existing accessibility label. Should I create a separate localized string for the `largeContentTitle`?

File: Client/Frontend/Strings.swift [hunk 11]
```diff
@@ -4701,6 +4706,21 @@ extension String {
         tableName: "TabLocation",
         value: "Secure connection. Enhanced Tracking Protection is off.",
         comment: "Accessibility label for the security icon in url bar")
+    public static let TabLocationLockButtonLargeContentTitle = MZLocalizedString(
+        key: "TabLocation.LockButton.LargeContentTitle.v122",
+        tableName: nil,
```
Comment (lmarceau) [inline Client/Frontend/Strings.swift:4711]: `TableName` shouldn't be nil for new strings. You can use `TabLocation` I think

File: Client/Frontend/Strings.swift [hunk 12]
```diff
@@ -4701,6 +4706,21 @@ extension String {
...
+        value: "Tracking Protection",
+        comment: "Large content title for the lock button")
```
Comment (lmarceau) [inline Client/Frontend/Strings.swift:4713]: This comment is used to provide a description for contributors who translate strings in [Pontoon](https://pontoon.mozilla.org/). I think we should give them a bit more context here as to what this string is used for i.e. for large content viewer accessibility feature on toolbar buttons.

File: Client/Frontend/Strings.swift [hunk 13]
```diff
@@ -5356,6 +5376,11 @@ extension String {
         tableName: nil,
         value: nil,
         comment: "Accessibility label for the tabs button in the (top) tab toolbar")
+    public static let TabsButtonShowTabsLargeContentTitle = MZLocalizedString(
```
Comment (lmarceau) [inline Client/Frontend/Strings.swift:5379]: I think you can put this string with the other ones above, using the same `tableName`. The string ordering in this file is a bit out of order when it's older strings.

File: Client/Frontend/Strings.swift [hunk 14]
```diff
@@ -4701,6 +4706,26 @@ extension String {
         tableName: "TabLocation",
         value: "Secure connection. Enhanced Tracking Protection is off.",
         comment: "Accessibility label for the security icon in url bar")
+    public static let TabLocationLockButtonLargeContentTitle = MZLocalizedString(
+        key: "TabLocation.LockButton.LargeContentTitle.v122",
+        tableName: "TabLocation",
+        value: "Tracking Protection",
+        comment: "Large content title for the lock button.  This title is displayed when using accessible font sizes is enabled")
```
Comment (lmarceau) [inline Client/Frontend/Strings.swift:4713]: nit, there's an extra space in all the comments
```suggestion
        comment: "Large content title for the lock button. This title is displayed when using accessible font sizes is enabled")
```

**Review by lmarceau — COMMENTED** [review]
One nit, otherwise LGTM!

**Comment by lmarceau** [comment]
Build failed due to this error. Could you look into this @atfelix ?

```
❌  /Users/vagrant/git/firefox-ios/firefox-ios-tests/Tests/ClientTests/TabToolbarHelperTests.swift:85:7: type 'MockTabToolbar' does not conform to protocol 'TabToolbarProtocol'
class MockTabToolbar: TabToolbarProtocol {
                                                     ^
❌  /Users/vagrant/git/firefox-ios/firefox-ios-tests/Tests/ClientTests/TabToolbarHelperTests.swift:85:7: type 'MockTabToolbar' does not conform to protocol 'TabToolbarProtocol'
class MockTabToolbar: TabToolbarProtocol {
      ^
```

**Comment by atfelix** [comment]
I'm looking into the potential memory leak indicated by `TabToolbarHelperTests.test_tabToolBarHelper_basicCreation_doesntLeak`

**Comment by atfelix** [comment]
> I'm looking into the potential memory leak indicated by `TabToolbarHelperTests.test_tabToolBarHelper_basicCreation_doesntLeak`

This issue should be resolved in the most recent commit

**Pull request merged**